IA-2588 Add zone and region to createRuntime requests

wnojopra opened the pull request:
Followup from #1920 (needed to change the branch name)

What I expect this PR does:
- Add an optional zone parameter to the createRuntime API
- Updates Swagger docs
- Adds a migration to add zone to the RUNTIME_CONFIG table
- Updates appropriate slick mapping
- Pass the zone is already passed to GCE. 

What I still need help with:
- Testing this locally. I think I have the right approach, but I still don't know how to check that what I'm doing will actually work?
- Testing the liquibase migration. How can I test this?

In general I am asking for feedback on if this is the right approach so far, and how I should proceed to make sure this change works. Thanks!
---
Have you read [CONTRIBUTING.md](https://github.com/DataBiosphere/leonardo/blob/develop/CONTRIBUTING.md) lately? If not, do that first.

I, the developer opening this PR, do solemnly pinky swear that:

- [x] I've documented my API changes in Swagger

In all cases:

- [ ] Get a thumbsworth of review and PO signoff if necessary
- [ ] Verify all tests go green
- [ ] Squash and merge; Delete your branch after this
- [ ] Test this change deployed correctly and works on dev environment after deployment

Commit:
Add zone to createRuntime requests

Persist zone to Leo database

Commit:
Add default zone to reference.conf

Commit:
Run scalafmt

Commit:
Make zone non optional. Update GCEInterpreter to use actual zone instead of config zone

Commit:
Run scalafmt

Commit:
Fix typos in test

Commit:
Fix typo in xml

Commit:
Fix failing pubsub tests

Commit:
Fix tests due to decode issues

Commit:
Add zone to another test

wnojopra committed:
Merge branch 'develop' into wn-zone

wnojopra commented:
jenkins retest

Qi77Qi commented on http/src/main/resources/org/broadinstitute/dsde/workbench/leonardo/liquibase/changesets/20210225_add_zone.xml:
```diff
@@ -0,0 +1,13 @@
+    <changeSet logicalFilePath="leonardo" author="willyn" id="add_zone">
+        <addColumn tableName="RUNTIME_CONFIG">
+            <column name="zone" type="varchar(254)">
+                <constraints nullable="true"/>
```
I don't think this should be `nullable`. we should populate existing rows with the old default zone, and make this non-nullable

Qi77Qi reviewed (commented):
once the migration I mentioned in the comment is done. I think ideally we should test this out in a fiab (altho this wouldn't be easy for you to set up, so maybe one of us can help), or you could add an automation test similar to [this](https://github.com/DataBiosphere/leonardo/blob/develop/automation/src/test/scala/org/broadinstitute/dsde/workbench/leonardo/runtimes/RuntimeCreationDiskSpec.scala#L39)

Commit:
Set nullable to false for zone column

rtitle commented on core/src/main/scala/org/broadinstitute/dsde/workbench/leonardo/http/runtimeRoutesModels.scala:
```diff
@@ -53,7 +55,8 @@ object RuntimeConfigRequest {
                                   numberOfPreemptibleWorkers: Option[Int] = None,
-                                  properties: Map[String, String])
+                                  properties: Map[String, String],
+                                  zone: Option[ZoneName] = None)
```
So Dataproc accepts a region _or_ a zone (unlike GCE which only accepts a zone). See: https://cloud.google.com/dataproc/docs/concepts/configuring-clusters/auto-zone

I'm wondering if we should have both in our API to cover both cases, e.g.:
```
region: Option[RegionName] = None,
zone: Option[ZoneName] = None
```

(And actually I don't see in this PR where the Dataproc configs are used in `DataprocInterpreter.scala`.)

rtitle replied:
Or maybe we _only_ support region for Dataproc

rtitle commented on http/src/main/resources/org/broadinstitute/dsde/workbench/leonardo/liquibase/changesets/20210225_add_zone.xml:
```diff
@@ -0,0 +1,13 @@
+    <changeSet logicalFilePath="leonardo" author="willyn" id="add_zone">
+        <addColumn tableName="RUNTIME_CONFIG">
+            <column name="zone" type="varchar(254)">
+                <constraints nullable="false"/>
```
If adding a new non-nullable field to an existing table it should be coupled with a data migration:
```
1. add nullable column
2. `UPDATE RUNTIME_CONFIG set zone = 'us-central1-a'`
3. Make column non-nullable

rtitle commented on http/src/main/resources/reference.conf:
```diff
@@ -31,6 +31,7 @@ dataproc {
     workerDiskSize = 100
     numberOfWorkerLocalSSDs = 0
     numberOfPreemptibleWorkers = 0
+    zone = "us-central1-a"
```
`zone` and `region` already exist up above.

This will actually change behavior: it will cause all Dataproc clusters to be created in the `us-central1-a` zone whereas currently they are created in the `us-central1` region (zones a through f). 

I'm thinking we should still use region for Dataproc clusters and zone for GCE.

rtitle replied:
Actually maybe `zone` and `region` on L15-16 should be removed in this PR now that they are tracked in the `RUNTIME_CONFIG` table.

rtitle commented on http/src/main/resources/reference.conf:
```diff
@@ -71,6 +72,7 @@ gce {
     machineType = "n1-standard-4"
     diskSize = 30 # This is default size for just user disk
     bootDiskSize = 50
+    zone = "us-central1-a"
```
Similar to above, probably can remove `region` and `zone` on L60-62

rtitle commented on http/src/main/scala/org/broadinstitute/dsde/workbench/leonardo/db/RuntimeConfigQueries.scala:
```diff
@@ -63,6 +63,12 @@ object RuntimeConfigQueries {
       .map(c => (c.machineType, c.dateAccessed))
       .update((machineType, dateAccessed))
 
+  def updateZone(id: RuntimeConfigId, zone: ZoneName, dateAccessed: Instant): DBIO[Int] =
```
Is this used?

rtitle commented on http/src/main/scala/org/broadinstitute/dsde/workbench/leonardo/util/GceInterpreter.scala:
```diff
@@ -69,9 +69,20 @@ class GceInterpreter[F[_]: Parallel: ContextShift](
         SetUpProjectFirewallsParams(params.runtimeProjectAndName.googleProject, network)
       )
 
+      zoneParam <- params.runtimeConfig match {
```
Changes in this file LGTM, but think we need equivalent in `DataprocInterpreter`.

Or, if we want to punt Dataproc regionality for this PR, remove it from the API/swagger page.

rtitle commented:
I left some comments -- mostly around Dataproc support. I think it's close though. Agree with Qi about testing but I think we can help with that since it relies on some Broad infrastructure like fiabs.

Again, thank you for contributing! This feature has been a long time coming, haha.

rtitle reviewed (changes requested):
see above

wnojopra commented:
Thanks, Rob! Initially the thought was to wire zone through just for GCE, but after thinking through your comments I think it makes sense to also plumb in the support for Dataproc in this same PR. I will be working on this today.

Commit:
Add data migration for zone field

Commit:
WIP of adding in region

Commit:
Add fix (?) for RuntimeConfigTable

wnojopra committed:
Merge branch 'develop' into wn-zone

Commit:
Use region instead of zone for dataproc in reference.conf

Commit:
Run scalafmt

Commit:
Update DataprocInterpreter

wnojopra commented:
Hi @rtitle , I've made updates to get region in dataproc supported, but I'm a bit stuck fixing a few compilation errors related to DataprocInterpreter.scala changes, notably this one:

`[error] /usr/local/google/home/willyn/repos/leonardo/http/src/main/scala/org/broadinstitute/dsde/workbench/leonardo/util/DataprocInterpreter.scala:261:9: type mismatch;
[error]  found   : _1 where type _1 >: org.broadinstitute.dsde.workbench.leonardo.RuntimeConfig.DataprocConfig with org.broadinstitute.dsde.workbench.google2.RegionName <: Product with java.io.Serializable
[error]  required: org.broadinstitute.dsde.workbench.google2.RegionName
[error]         regionParam,
[error]         ^
`

I _think_ I'm pattern matching what I'm doing in GCEInterpreter, so I could use some help understanding this error. Thanks!

rtitle commented:
Hey @wnojopra, will take a look! Do you mind if I push some commits to this PR?

wnojopra commented:
> Hey @wnojopra, will take a look! Do you mind if I push some commits to this PR?

Please, go ahead! I appreciate it.

rtitle committed:
Fix-up:
- make RUNTIME_CONFIG.[zone|region] nullable in the DB because we expect either to be present but not both
- remove old zone/region configs, use RuntimeConfig fields for all monitoring
- some general code cleanup in these areas

rtitle committed:
Merge remote-tracking branch 'origin/develop' into wn-zone

rtitle committed:
Fix merge issues

rtitle committed:
Merge remote-tracking branch 'origin/develop' into wn-zone

rtitle committed:
Fix broken unit tests

rtitle commented:
jenkins retest

Commit:
Add runtime tests for zone/region parameters

Commit:
scalafmt

wnojopra commented:
jenkins retest

wnojopra committed:
Merge branch 'develop' into wn-zone

wnojopra committed:
Merge branch 'develop' into wn-zone

Commit:
Add RuntimeGceSpec test to Leonardo suite

rtitle commented on automation/src/test/scala/org/broadinstitute/dsde/workbench/leonardo/runtimes/RuntimeGceSpec.scala:
```diff
@@ -0,0 +1,53 @@
+        _ <- LeonardoApiClient.deleteRuntime(project, runtimeName)
+      } yield ()
+    }
+  }
```
You need `res.unsafeRunSync()` to actually run this test.

rtitle replied:
(added it)

rtitle commented:
The new tests look good (one comment above).

For some reason Dataproc tests are failing. Here's a link to the test run: https://fc-jenkins.dsp-techops.broadinstitute.org/job/leonardo-fiab-test-runner/34472/

I see this in Leo logs:
```
{
   "@timestamp":"2021-03-25T20:24:48.093Z",
   "@version":"1",
   "message":"{\"response\":null,\"result\":\"Failed\"}",
   "logger_name":"org.broadinstitute.dsde.workbench.leonardo.http.Boot",
   "thread_name":"ioapp-compute-2",
   "severity":"ERROR",
   "level_value":40000,
   "stack_trace":"com.google.api.gax.rpc.UnimplementedException: io.grpc.StatusRuntimeException: UNIMPLEMENTED: HTTP status code 404\ninvalid content-type: text/html; charset=UTF-8\nheaders: Metadata(:status=404,content-type=text/html; charset=UTF-8,referrer-policy=no-referrer,content-length=1617,date=Thu, 25 Mar 2021 20:24:48 GMT)\nDATA-----------------------------\n<!DOCTYPE html>\n<html lang=en>\n  <meta charset=utf-8>\n  <meta name=viewport content=\"initial-scale=1, minimum-scale=1, width=device-width\">\n  <title>Error 404 (Not Found)!!1</title>\n  <style>\n    *{margin:0;padding:0}html,code{font:15px/22px arial,sans-serif}html{background:#fff;color:#222;padding:15px}body{margin:7% auto 0;max-width:390px;min-height:180px;padding:30px 0 15px}* > body{background:url(//www.google.com/images/errors/robot.png) 100% 5px no-repeat;padding-right:205px}p{margin:11px 0 22px;overflow:hidden}ins{color:#777;text-decoration:none}a img{border:0}@media screen and (max-width:772px){body{background:none;margin-top:0;max-width:none;padding-right:0}}#logo{background:url(//www.google.com/images/branding/googlelogo/1x/googlelogo_color_150x54dp.png) no-repeat;margin-left:-5px}@media only screen and (min-resolution:192dpi){#logo{background:url(//www.google.com/images/branding/googlelogo/2x/googlelogo_color_150x54dp.png) no-repeat 0% 0%/100% 100%;-moz-border-image:url(//www.google.com/images/branding/googlelogo/2x/googlelogo_color_150x54dp.png) 0}}@media only screen and (-webkit-min-device-pixel-ratio:2){#logo{background:url(//www.google.com/images/branding/googlelogo/2x/googlelogo_color_150x54dp.png) no-repeat;-webkit-background-size:100% 100%}}#logo{display:inline-block;height:54px;width:150px}\n  </style>\n  <a href=//www.google.com/><span id=logo aria-label=Google></span></a>\n  <p><b>404.</b> <ins>Thatâ€™s an error.</ins>\n  <p>The requested URL <code>/google.cloud.dataproc.v1.ClusterController/CreateCluster</code> was not found on this server.  <ins>Thatâ€™s all we know.</ins>\n\n\tat com.google.api.gax.rpc.ApiExceptionFactory.createException(ApiExceptionFactory.java:65)\n\tat com.google.api.gax.grpc.GrpcApiExceptionFactory.create(GrpcApiExceptionFactory.java:72)\n\tat com.google.api.gax.grpc.GrpcApiExceptionFactory.create(GrpcApiExceptionFactory.java:60)\n\tat com.google.api.gax.grpc.GrpcExceptionCallable$ExceptionTransformingFuture.onFailure(GrpcExceptionCallable.java:97)\n\tat com.google.api.core.ApiFutures$1.onFailure(ApiFutures.java:68)\n\tat com.google.common.util.concurrent.Futures$CallbackListener.run(Futures.java:1074)\n\tat com.google.common.util.concurrent.DirectExecutor.execute(DirectExecutor.java:30)\n\tat com.google.common.util.concurrent.AbstractFuture.executeListener(AbstractFuture.java:1213)\n\tat com.google.common.util.concurrent.AbstractFuture.complete(AbstractFuture.java:983)\n\tat com.google.common.util.concurrent.AbstractFuture.setException(AbstractFuture.java:771)\n\tat io.grpc.stub.ClientCalls$GrpcFuture.setException(ClientCalls.java:563)\n\tat io.grpc.stub.ClientCalls$UnaryStreamToFuture.onClose(ClientCalls.java:533)\n\tat io.grpc.internal.DelayedClientCall$DelayedListener$3.run(DelayedClientCall.java:464)\n\tat io.grpc.internal.DelayedClientCall$DelayedListener.delayOrExecute(DelayedClientCall.java:428)\n\tat io.grpc.internal.DelayedClientCall$DelayedListener.onClose(DelayedClientCall.java:461)\n\tat io.grpc.internal.ClientCallImpl.closeObserver(ClientCallImpl.java:553)\n\tat io.grpc.internal.ClientCallImpl.access$300(ClientCallImpl.java:68)\n\tat io.grpc.internal.ClientCallImpl$ClientStreamListenerImpl$1StreamClosed.runInternal(ClientCallImpl.java:739)\n\tat io.grpc.internal.ClientCallImpl$ClientStreamListenerImpl$1StreamClosed.runInContext(ClientCallImpl.java:718)\n\tat io.grpc.internal.ContextRunnable.run(ContextRunnable.java:37)\n\tat io.grpc.internal.SerializingExecutor.run(SerializingExecutor.java:123)\n\tat java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:511)\n\tat java.util.concurrent.FutureTask.run(FutureTask.java:266)\n\tat java.util.concurrent.ScheduledThreadPoolExecutor$ScheduledFutureTask.access$201(ScheduledThreadPoolExecutor.java:180)\n\tat java.util.concurrent.ScheduledThreadPoolExecutor$ScheduledFutureTask.run(ScheduledThreadPoolExecutor.java:293)\n\tat java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1149)\n\tat java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:624)\n\tat java.lang.Thread.run(Thread.java:748)\n\tat async @ org.broadinstitute.dsde.workbench.google2.GoogleDataprocInterpreter.$anonfun$createCluster$6(GoogleDataprocInterpreter.scala:69)\n\tat map @ org.broadinstitute.dsde.workbench.google2.GoogleDataprocInterpreter.$anonfun$createCluster$6(GoogleDataprocInterpreter.scala:69)\n\tat flatMap @ org.broadinstitute.dsde.workbench.google2.GoogleDataprocInterpreter.createCluster(GoogleDataprocInterpreter.scala:68)\n\tat delay$extension @ fs2.io.file.FileHandle$.fromPath(FileHandle.scala:107)\n\tat apply @ org.broadinstitute.dsde.workbench.leonardo.http.Boot$.$anonfun$createDependencies$1(Boot.scala:334)\n\tat withPermit @ org.http4s.client.PoolManager.shutdown(PoolManager.scala:370)\n\tat apply @ org.broadinstitute.dsde.workbench.leonardo.http.Boot$.$anonfun$createDependencies$1(Boot.scala:334)\n\tat withPermit @ org.http4s.client.PoolManager.shutdown(PoolManager.scala:370)\n\tat flatMap @ org.broadinstitute.dsde.workbench.google2.package$.$anonfun$withLogging$1(package.scala:64)\n\tat flatMap @ org.broadinstitute.dsde.workbench.google2.package$.withLogging(package.scala:63)\n\tat map @ org.broadinstitute.dsde.workbench.google2.package$.$anonfun$tracedLogging$1(package.scala:99)\n\tat map @ org.broadinstitute.dsde.workbench.leonardo.package$$anon$1.ask(package.scala:29)\n\tat flatMap @ org.broadinstitute.dsde.workbench.google2.package$.tracedLogging(package.scala:98)\nCaused by: io.grpc.StatusRuntimeException: UNIMPLEMENTED: HTTP status code 404\ninvalid content-type: text/html; charset=UTF-8\nheaders: Metadata(:status=404,content-type=text/html; charset=UTF-8,referrer-policy=no-referrer,content-length=1617,date=Thu, 25 Mar 2021 20:24:48 GMT)\nDATA-----------------------------\n<!DOCTYPE html>\n<html lang=en>\n  <meta charset=utf-8>\n  <meta name=viewport content=\"initial-scale=1, minimum-scale=1, width=device-width\">\n  <title>Error 404 (Not Found)!!1</title>\n  <style>\n    *{margin:0;padding:0}html,code{font:15px/22px arial,sans-serif}html{background:#fff;color:#222;padding:15px}body{margin:7% auto 0;max-width:390px;min-height:180px;padding:30px 0 15px}* > body{background:url(//www.google.com/images/errors/robot.png) 100% 5px no-repeat;padding-right:205px}p{margin:11px 0 22px;overflow:hidden}ins{color:#777;text-decoration:none}a img{border:0}@media screen and (max-width:772px){body{background:none;margin-top:0;max-width:none;padding-right:0}}#logo{background:url(//www.google.com/images/branding/googlelogo/1x/googlelogo_color_150x54dp.png) no-repeat;margin-left:-5px}@media only screen and (min-resolution:192dpi){#logo{background:url(//www.google.com/images/branding/googlelogo/2x/googlelogo_color_150x54dp.png) no-repeat 0% 0%/100% 100%;-moz-border-image:url(//www.google.com/images/branding/googlelogo/2x/googlelogo_color_150x54dp.png) 0}}@media only screen and (-webkit-min-device-pixel-ratio:2){#logo{background:url(//www.google.com/images/branding/googlelogo/2x/googlelogo_color_150x54dp.png) no-repeat;-webkit-background-size:100% 100%}}#logo{display:inline-block;height:54px;width:150px}\n  </style>\n  <a href=//www.google.com/><span id=logo aria-label=Google></span></a>\n  <p><b>404.</b> <ins>Thatâ€™s an error.</ins>\n  <p>The requested URL <code>/google.cloud.dataproc.v1.ClusterController/CreateCluster</code> was not found on this server.  <ins>Thatâ€™s all we know.</ins>\n\n\tat io.grpc.Status.asRuntimeException(Status.java:535)\n\tat io.grpc.stub.ClientCalls$UnaryStreamToFuture.onClose(ClientCalls.java:533)\n\tat io.grpc.internal.DelayedClientCall$DelayedListener$3.run(DelayedClientCall.java:464)\n\tat io.grpc.internal.DelayedClientCall$DelayedListener.delayOrExecute(DelayedClientCall.java:428)\n\tat io.grpc.internal.DelayedClientCall$DelayedListener.onClose(DelayedClientCall.java:461)\n\tat io.grpc.internal.ClientCallImpl.closeObserver(ClientCallImpl.java:553)\n\tat io.grpc.internal.ClientCallImpl.access$300(ClientCallImpl.java:68)\n\tat io.grpc.internal.ClientCallImpl$ClientStreamListenerImpl$1StreamClosed.runInternal(ClientCallImpl.java:739)\n\tat io.grpc.internal.ClientCallImpl$ClientStreamListenerImpl$1StreamClosed.runInContext(ClientCallImpl.java:718)\n\tat io.grpc.internal.ContextRunnable.run(ContextRunnable.java:37)\n\tat io.grpc.internal.SerializingExecutor.run(SerializingExecutor.java:123)\n\tat java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:511)\n\tat java.util.concurrent.FutureTask.run(FutureTask.java:266)\n\tat java.util.concurrent.ScheduledThreadPoolExecutor$ScheduledFutureTask.access$201(ScheduledThreadPoolExecutor.java:180)\n\tat java.util.concurrent.ScheduledThreadPoolExecutor$ScheduledFutureTask.run(ScheduledThreadPoolExecutor.java:293)\n\tat java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1149)\n\tat java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:624)\n\tat java.lang.Thread.run(Thread.java:748)\n",
   "traceId":"2e22bb03-eac9-43d1-a393-2d08e2762d84",
   "googleCall":"com.google.cloud.dataproc.v1.ClusterControllerClient.createClusterAsync(RegionName(us-central1), DataprocClusterName(automation-test-abwsbthrz), Some(CreateClusterConfig(service_account_scopes: \"https://www.googleapis.com/auth/devstorage.full_control\"\nservice_account_scopes: \"https://www.googleapis.com/auth/bigquery\"\nservice_account_scopes: \"https://www.googleapis.com/auth/logging.write\"\nservice_account_scopes: \"https://www.googleapis.com/auth/userinfo.email\"\nservice_account_scopes: \"https://www.googleapis.com/auth/source.read_only\"\nservice_account_scopes: \"https://www.googleapis.com/auth/cloud.useraccounts.readonly\"\nservice_account_scopes: \"https://www.googleapis.com/auth/userinfo.profile\"\ntags: \"leonardo\"\nsubnetwork_uri: \"leonardo-subnetwork\"\nservice_account: \"b044pet-109021626430559047321@gpalloc-qa-master-fhix5rt.iam.gserviceaccount.com\"\n,List(executable_file: \"gs://leoinit-automation-test-ab-fb60ca8f-8c78-4253-abea-6957d29097e8/init-actions.sh\"\nexecution_timeout {\n  seconds: 1800\n}\n),image_uri: \"projects/broad-dsp-gcr-public/global/images/custom-leo-image-dataproc-1-4-51-debian10-d26876f\"\nmachine_type_uri: \"n1-standard-4\"\ndisk_config {\n  boot_disk_size_gb: 100\n}\n,Some(num_instances: 2\nimage_uri: \"projects/broad-dsp-gcr-public/global/images/custom-leo-image-dataproc-1-4-51-debian10-d26876f\"\nmachine_type_uri: \"n1-standard-4\"\ndisk_config {\n  boot_disk_size_gb: 100\n}\n),Some(num_instances: 1\nimage_uri: \"projects/broad-dsp-gcr-public/global/images/custom-leo-image-dataproc-1-4-51-debian10-d26876f\"\nmachine_type_uri: \"n1-standard-4\"\ndisk_config {\n  boot_disk_size_gb: 100\n}\nis_preemptible: true\n),leostaging-automation-test-206fa43c-e941-4970-82c7-8d19ef7a4163,properties {\n  key: \"dataproc:dataproc.monitoring.stackdriver.enable\"\n  value: \"true\"\n}\nproperties {\n  key: \"spark:spark.hadoop.fs.gs.requester.pays.mode\"\n  value: \"AUTO\"\n}\nproperties {\n  key: \"spark:spark.hadoop.fs.gs.requester.pays.project.id\"\n  value: \"gpalloc-qa-master-fhix5rt\"\n}\nproperties {\n  key: \"yarn:yarn.log-aggregation-enable\"\n  value: \"true\"\n}\n))))",
   "duration":"44 milliseconds",
   "result":"Failed"
}
```

The error looks weird to me, looks like the Dataproc API call is returning 404? The fields look correct to me, and the region is `us-central1` in this case. Can't see in the code where the bug might be either. Will look some more tomorrow.

rtitle committed:
Add unsafeRunSync

rtitle committed:
Fix RuntimeGceSpec

rtitle commented:
OK, I think this is why Dataproc is failing. Currently in `Boot.scala` we're passing `RegionName("unused")` to the `GoogleDataprocService` constructor. I thought this wasn't used but looks like it is: https://github.com/broadinstitute/workbench-libs/blob/develop/google2/src/main/scala/org/broadinstitute/dsde/workbench/google2/GoogleDataprocService.scala#L109

Does the Dataproc Java lib require a new client per-region? 🤦  I wonder if there's a way to make a single Dataproc client work with multiple regions. @wnojopra do you want to look into that? Feel free to PR to workbench-libs if needed.

I thought of a couple more things too. We probably need to make disks (https://github.com/DataBiosphere/leonardo/blob/develop/http/src/main/resources/reference.conf#L573) and VPC (https://github.com/DataBiosphere/leonardo/blob/develop/http/src/main/resources/reference.conf#L112) multi-zone as well to support instances in multiple zones. Can think about this more… but I think scoping this PR to GCE and Dataproc makes sense, and we can follow up in a separate PR about on those things. (Though I wonder if the tests will pass as is…)

wnojopra commented:
Thanks @rtitle. I'm not familiar with the Java lib but I'll look into it. Sounds like something we'll have to figure out. 

I think as long as we make the tests pass we can follow up with VPC and multi-zone in a future PR.

wnojopra commented:
@rtitle I think we need one client per-region. 

I did try to play around with the [Global Endpoint](https://cloud.google.com/dataproc/docs/concepts/regional-endpoints) a bit, but this doesn't seem to do what we'd like. It'd be nice if it worked as a global endpoint that can make clusters in any region. What I think it actually does is allow you to make clusters in any zone. And worst of all, its mentioned as to-be deprecated.

I don't think having one client per-region is so bad. What I propose is: in Boot.scala, instead of setting dataprocService to be a GoogleDataprocService, it is instead a Map of GoogleDataprocServices, keyed by RegionName. Then we'd change all usages of this dataprocService object to look up by region first. There shouldn't be much overhead here because this is all client side setup. What do you think?

rtitle commented:
Yeah, something like that sounds good. Could probably even just instantiate a new `GoogleDataprocService` each time one is needed.

Commit:
Make Leonardo's dataprocService a map of services keyed by region

wnojopra committed:
Merge branch 'develop' into wn-zone

Commit:
scalafmt

Commit:
Fix tests by using us-central1 region

Commit:
scalafmt

Commit:
Fix for broken europe-west1 test

rtitle committed:
1. Use configured Map of Dataproc regions and make handling more FP friendly
2. Attempt to fix RuntimeDataprocSpec by using a region-specific Dataproc client

Commit:
Fix typos for encoder names. Add more regions to config

wnojopra committed:
Merge branch 'develop' into wn-zone

wnojopra commented on automation/src/test/scala/org/broadinstitute/dsde/workbench/leonardo/LeonardoTestUtils.scala:
```diff
@@ -97,13 +96,21 @@ trait LeonardoTestUtils
+            compute,
+            LeonardoConfig.GCS.pathToQAJson,
+            blocker,
+            semaphore,
```
@rtitle I notice the other semaphores in this file use unsafeRunSync. Is it ok to use just plain `semaphore` here? Where does that even come from?

rtitle replied:
It comes from the `LeonardoTestSuite` super-trait: https://github.com/DataBiosphere/leonardo/blob/develop/core/src/test/scala/org/broadinstitute/dsde/workbench/leonardo/LeonardoTestSuite.scala#L27

I think this is fine, I think we just put `semaphore` in the trait to avoid duplication, but didn't update everywhere yet. Actually it would probably be better for all these clients to use the same semaphore, instead of creating a new one each time.

rtitle committed:
Fix encoding/decoding issues

rtitle commented:
@wnojopra I _think_ I just fixed the encoding/decoding issues in commit https://github.com/DataBiosphere/leonardo/pull/1933/commits/ce1eb6b2b4f697ab2bb15f101fcb1646d3f656ac. Explanation inline below…

rtitle commented on core/src/main/scala/org/broadinstitute/dsde/workbench/leonardo/http/runtimeRoutesModels.scala:
```diff
@@ -34,7 +34,6 @@ object RuntimeConfigRequest {
   final case class GceConfig(
     machineType: Option[MachineTypeName],
     diskSize: Option[DiskSize],
-    region: Option[ZoneName] = None,
```
This was subtle: an extraneous `region` param in `RuntimeConfigRequest.GceConfig`. Because both `ZoneName` and `RegionName` encode from strings, and each has default arguments, I think circe was getting confused and populating `region` instead of `zone`, which had no effect.

rtitle commented on core/src/test/scala/org/broadinstitute/dsde/workbench/leonardo/http/RuntimeRoutesTestJsonCodec.scala:
```diff
@@ -23,7 +23,7 @@ import java.time.Instant
 
 object RuntimeRoutesTestJsonCodec {
-  implicit val dataprocConfigEncoder: Encoder[RuntimeConfigRequest.DataprocConfig] = Encoder.forProduct9(
+  implicit val dataprocConfigRequestEncoder: Encoder[RuntimeConfigRequest.DataprocConfig] = Encoder.forProduct9(
```
Renamed some of these implicits to not name clash, mostly for the unit test in `HttpRoutesSpec`

rtitle commented on http/src/main/scala/org/broadinstitute/dsde/workbench/leonardo/http/api/RuntimeRoutes.scala:
```diff
@@ -314,7 +314,8 @@ object RuntimeRoutes {
                                                 numberOfPreemptibleWorkers,
-                                                properties)
+                                                properties,
+                                                region)
```
Bug number 2: `region` was not being passed to `RuntimeConfigRequest.DataprocConfig` here (causing it to use the default region always).

rtitle commented on http/src/test/scala/org/broadinstitute/dsde/workbench/leonardo/http/api/HttpRoutesSpec.scala:
```diff
@@ -556,6 +558,41 @@ class HttpRoutesSpec
     }
   }
 
+  // Validate encoding/decoding of RuntimeConfigRequest types
```
Added some tests that round-trip encode/decode these types.

I put it here because the `RuntimeConfigRequest.*` decoders are defined in the `http` module in `RuntimeRoutes`. I could see moving this to `JsonCodec`/`JsonCodecSpec` in `core`, but that would be a bit more refactoring.

rtitle committed:
Fix compilation error

rtitle committed:
Merge branch 'develop' into wn-zone

Commit:
Change europe-west1-a to europe-west1-b. Apparently europe-west1-a doesn't exist!

rtitle commented:
OK, now it's actually trying to create the runtimes in the other region, but failing with:

```
400 Bad Request
POST https://compute.googleapis.com/compute/v1/projects/gpalloc-qa-master-tsxdnkc/zones/europe-west1-b/instances
{
  "error": {
    "code": 400,
    "message": "Invalid value for field 'resource.networkInterfaces[0].subnetwork': 'projects/gpalloc-qa-master-tsxdnkc/regions/us-central1/subnetworks/leonardo-subnetwork'. Instance must be in the same region as the subnetwork.",
    "errors": [
      {
        "message": "Invalid value for field 'resource.networkInterfaces[0].subnetwork': 'projects/gpalloc-qa-master-tsxdnkc/regions/us-central1/subnetworks/leonardo-subnetwork'. Instance must be in the same region as the subnetwork.",
        "domain": "global",
        "reason": "invalid"
      }
    ]
  }
}
```

I think it's because of this: https://github.com/DataBiosphere/leonardo/blob/develop/http/src/main/resources/reference.conf#L112. E.g. we always create the instance's subnet in the `us-central1` region.

Thinking about the approach, we _could_ try to use global mode for VPC creation (see [considerations](https://cloud.google.com/vpc/docs/vpc#auto-mode-considerations) here). But I think we didn't want to do that previously because of (a) unnecessary subnets created in regions we're not using; and (b) need to use less restrictive firewall because GCP picks the subnet IP ranges for us.

Maybe a better approach is to have [VPCInterpreter](https://github.com/DataBiosphere/leonardo/blob/develop/http/src/main/scala/org/broadinstitute/dsde/workbench/leonardo/util/VPCInterpreter.scala#L167) not use the configured region but simply accept `region` as a parameter in [SetUpProjectNetworkParams](https://github.com/DataBiosphere/leonardo/blob/develop/http/src/main/scala/org/broadinstitute/dsde/workbench/leonardo/util/VPCInterpreter.scala#L61). That way we set up the subnet in the region we're using for the particular runtime request.

(BTW, we have a ticket to move all this VPC setup stuff outside of Leo, and use Resource Buffer Service instead.)

Qi77Qi committed:
use multi-region supported dataproc client

Qi77Qi commented:
@wnojopra I'm going to make some updates. specificaly, I made [this pr](https://github.com/broadinstitute/workbench-libs/pull/570) so that dataproc client can support multiple regions. I'll try to incorporate the new library in this PR...lmk if you'd like me to hold

wnojopra commented:
@Qi77Qi Please feel free, that would be great! I'm also working on fixes to the subnetwork creation but I don't think that will conflict with your work.

Commit:
Pass in a region parameter to subnetwork creation

Qi77Qi committed:
use published wb-libs

Commit:
Instead of a set IP range, add a map from region to IP range in the config

Commit:
Add us-west to fix unit test

wnojopra committed:
Merge branch 'develop' into wn-zone

Commit:
Try adding more sourceRanges into firewall config

wnojopra committed:
Merge branch 'develop' into wn-zone

wnojopra commented:
jenkins retest

Commit:
Use a map of regions -> sourceIpRange in config for firewall

Commit:
Merge branch 'wn-zone' of github.com:DataBiosphere/leonardo into wn-zone

wnojopra committed:
Merge branch 'develop' into wn-zone

Commit:
Fix dependencies from merge conflict

Commit:
Fix new unit tests

Qi77Qi committed:
remove unnecessary config reader code

Qi77Qi committed:
fixed firewall rule creation bug

Qi77Qi committed:
Merge branch 'develop' into wn-zone

Qi77Qi committed:
Merge branch 'develop' into wn-zone

Qi77Qi commented on http/src/main/scala/org/broadinstitute/dsde/workbench/leonardo/util/VPCInterpreter.scala:
```diff
@@ -190,4 +216,7 @@ final class VPCInterpreter[F[_]: Parallel: ContextShift: StructuredLogger: Timer
       )
       .build
+
+  private def buildFirewallName(prefix: String, regionName: RegionName): FirewallRuleName =
```
make firewall name region specific becuz subnets are region specific now

Qi77Qi reviewed (approved):
think this should be good to go

rtitle commented:
Whooa automation tests passed!! 🎉 

Will take a final look at this PR.

rtitle reviewed (approved):
🚀

Qi77Qi commented:
merging this!

Qi77Qi commented:
@wnojopra thanks for the great contribution...created [this ticket](https://broadworkbench.atlassian.net/browse/IA-2640) to track the work that needed to add zone support in diskRoutes API...there will also need UI work in order for this to be available for users. But this is a huge step towards supporting multi-zone and multi-regions!
